lib/consensus/ballot_send_record.go

```diff
@@ -0,0 +1,64 @@
+package consensus
+
+import (
+	"sync"
+
+	logging "github.com/inconshreveable/log15"
+)
+
+// Record the ballot sent by ISAACstate
+// This is to avoid sending another voting result in the same ISAACState.
+type BallotSendRecord struct {
+	sync.RWMutex
+
+	record map[ISAACState]bool
+	log    logging.Logger
+}
+
+func NewBallotSendRecord(nodeAlias string) *BallotSendRecord {
+	p := &BallotSendRecord{
+		record: make(map[ISAACState]bool),
+		log:    log.New(logging.Ctx{"node": nodeAlias}),
+	}
+
+	return p
+}
+
+func (r *BallotSendRecord) SetSent(state ISAACState) {
+	r.Lock()
+	defer r.Unlock()
+	log.Debug("BallotSendRecord.SetSent()", "state", state)
+	r.record[state] = true
+}
+
+func (r *BallotSendRecord) Sent(state ISAACState) bool {
+	r.RLock()
+	defer r.RUnlock()
+	log.Debug("BallotSendRecord.Sent()", "state", state)
+	sent, ok := r.record[state]
+	if !ok {
+		log.Debug("return false; state is not exist in BallotSendRecord.Sent", "state", state)
+		return false
+	}
+	return sent
+}
+
+func (r *BallotSendRecord) RemoveLowerThanOrEqualHeight(height uint64) {
+	r.Lock()
+	defer r.Unlock()
+	log.Debug("BallotSendRecord.RemoveLowerThanOrEqualHeight()", "height", height)
+
+	record := []ISAACState{}
```

Comment:
What do you make record?  Is it not OK ?
```
for s := range r.record {
   if s.Height <= height {
     delete(r.secord,s)
   }
}
```

Reply:
@anarcher Oh, It works! I'm worried about deleting in for loop.